### PR TITLE
Adds support for _.clone(DataView)

### PR DESCRIPTION
**Pull request by justsml**
I feel this might need better test coverage for both NodeJS & Browser... Any suggestions?

### CODE REVIEW

File: lodash.js [hunk 1]
```diff
@@ -3958,11 +3960,26 @@
...
+
+    /**
+     * Creates a clone of `dataView`.
+     *
+     * @private
+     * @param {DataView} dataView The data view to clone.
+     * @returns {DataView} Returns the cloned data view.
+     */
+    function cloneDataView(dataView) {
+      var buf = cloneArrayBuffer(dataView.buffer, dataView.byteOffset),
```
Comment (jdalton) [inline lodash.js]: Instead of cloning the `byteOffset` of the buffer wouldn't you clone the buffer whole and then provide the offset as:
```js
new dataView.constructor(buf, dataView.byteOffset, dataView.byteLength);
```

File: lodash.js [hunk 2]
```diff
@@ -85,6 +85,7 @@
       weakSetTag = '[object WeakSet]';
```
Comment (jdalton) [inline lodash.js]: You should add `DataView` to the `contextProps` array too.

**Comment by jdalton** [comment]
> I feel this might need better test coverage for both NodeJS & Browser... Any suggestions?

Ya, but don't worry about that. I'll take that bit on.

File: lodash.js (hunk 1 above)
Comment (justsml) [inline lodash.js]: My thought was it'd be less memory usage. 
I may be wrong, I haven't done much profiling with these parts of JS.

Comment (jdalton) [inline lodash.js]: I've punted on optimized routes for now (because its such an edge case at the moment).

File: test/test.js [hunk 3]
```diff
@@ -2634,6 +2636,20 @@
...
       });
 
+      QUnit.test('`_.' + methodName + '` should clone dataviews', function(assert) {
+        assert.expect(2);
+
+        if (DataView) {
+          var actual = func(dataView);
+
+          assert.strictEqual(actual.byteLength, dataView.byteLength);
+          assert.notStrictEqual(actual, dataView);
```
Comment (megawac) [inline test/test.js]: can you also check that it's  a `DataView` instance

**Comment by justsml** [comment]
I am assuming that a cloned DataView should re-create the underlying view-able buffer, and then return a new DataView w/ the byteOffset now 'reset' to 0 with the new buffer as it's contents.

Am I making too many assumptions?

File: lodash.js (hunk 1 above)
Comment (justsml) [inline lodash.js]: Also at first I thought I'd just make `cloneArrayBuffer` use `TypedArray.slice` - it seemed like an efficient way to get the data. 
@jdalton  I am guessing it isn't used because of IE10 support?

File: lodash.js (hunk 2 above)
Comment (justsml) [inline lodash.js]: Added :heavy_check_mark:

**Comment by justsml** [comment]
My thoughts on how this should behave are:

* Considering `DataView` doesn't contain much more than a couple integers (roughly speaking):
    1. **Pointer to memory addr** - potentially referencing a huge ArrayBuffer
    1. **Integer byteOffset** - which might limit what the DataView sees to, let's say only 8 bytes, out of perhaps many MBs.
    1. **Integer Length** - which reminds me, I forgot to account for this :astonished: 

* Would it be better to rely on `clone vs. cloneDeep` to achieve something like:
    * `_.cloneDeep(dv)` would copy the underlying buffer, resetting the offset & length? 
    * `_.clone(dv)` would simply `return new DataView(dv.buffer, dv.byteOffset, dv.byteLength)`. I think this seems a little pointless though - and it may not be what a typical user would expect?

File: lodash.js (hunk 1 above)
Comment (jdalton) [inline lodash.js]: Correct. Older-modernish environments don't support it.

**Comment by jdalton** [comment]
No worries. You've done a lot of work here. I can take it the rest of the way. The `DataView` is just another view like a typed array so can reuse the `cloneTypedArray` method (which I'll rename to `cloneArrayView`. The idea is if it's a shallow clone then just the view is cloned leaving the original array buffer in place. If it's a deep clone then the view and the buffer are cloned. No need to muck with resetting offsets we use the values of the original view being cloned.

**Comment by justsml** [comment]
@jdalton After thinking about it more, cloning a `view` type seems pointless - because you can't do:
```js
// !!!! Non-example/pattern
var section1 = _.clone(dv);
section1.byteOffset += 1024;
var section2 = _.clone(section1);
section2.byteOffset += 1024;
```

I can't think of a real use-case for shallow cloning a `DataView`
If I want to clone the buffer, I'll do that.

But if I want to clone a DataView, **surely I mean to get a new DataView w/ a copy of the visible buffer?**

```js
// so in code-words:
var dv1 = _.clone(someDataView);
// Is almost no different than doing this:
var dv1 = someDataView;
// DataView members are immutable after all
```

**Comment by jdalton** [comment]
Closed by https://github.com/lodash/lodash/commit/f6d8036a5e7a4b0945f79b0a021bf492dcfeaf96.

**Comment by justsml** [comment]
@jdalton Did my last msg make sense? Or am I missing something here?

**Comment by jdalton** [comment]
> Did my last msg make sense? Or am I missing something here?

Shallow cloning has limited value in some cases but it's a shallow clone nonetheless.

**Comment by justsml** [comment]
Ok. Uniformity makes for a good argument.

For my presumptive use-case, to extract ONLY the data presented by the DataView, should something like this be possible:
```js
var freshView = _.castArray(dataView);
```
